### PR TITLE
Use jinja2 when populating metadata maps

### PR DESCRIPTION
**Pull request by manics**
# What this PR does

The IDR bulk-annotation -> map-annotations process involves building URLs. Previously these were simple identifiers, but now they may include phrases including characters which need to be escaped when used in a URL. This PR now uses the Jinja templating module if it's installed, so standard Jinja2 filters such as `{{ value | urlencode }}` can be used.
# Testing this PR

Current use case: Mapr @aleksandra-tarkowska which creates mapr URLs with spaces.
# Potential issues

This adds a new optional dependency, `jinja2`, to OMERO.py. Should it be made mandatory along with yaml in order to keep things clearer?

### CODE REVIEW

File: components/tools/OmeroPy/src/omero/util/metadata_utils.py [hunk 1]
```diff
@@ -30,6 +30,13 @@
 from omero.constants import namespaces
 import re
 
+# TODO: Make jinja2 mandatory?
```
Comment (atarkowska) [inline components/tools/OmeroPy/src/omero/util/metadata_utils.py:33]: minimum version jinja2.7 http://jinja.pocoo.org/docs/dev/templates/#urlencode

Comment (atarkowska) [reply components/tools/OmeroPy/src/omero/util/metadata_utils.py:33]: might be good idea to add that as mandatory, otherwise you will ends up with:

```
Key: some URL
Value: https://someurl?key={{ value|urlencode }}
```

**Comment by atarkowska** [comment]
tested via http://10.0.51.150/web/mapr/compound/NIFLUMIC%20ACID/

Thank you for a prompt help, looks good, works as expected

**Comment by manics** [comment]
@aleksandra-tarkowska How's this?

**Comment by atarkowska** [comment]
by running:

```
bash-4.2$ bin/omero metadata populate --context bulkmap --cfg /tmp/idr-metadata/idr0016-wawer-bioactivecompoundprofiling/screenA/idr0016-screenA-bulkmap-config.yml Screen:1251
Using session c8faf254-c2a0-43d5-9549-142f1874f137 (demo@localhost:4064). Idle timeout: 10 min. Current group: demo
Traceback (most recent call last):
  File "bin/omero", line 125, in <module>
...
    raise JINJA2_MISSING
ImportError: No module named jinja2
```

**Comment by aleksandra-tarkowska** [comment]
although urlencode doesnt work anymore

```
idr-20160803=# select * from annotation_mapvalue where value like 'NIFLUMIC%';
 annotation_id |     name      |     value     | index 
---------------+---------------+---------------+-------
       2828971 | Compound Name | NIFLUMIC ACID |     0
(1 row)
```

**Comment by atarkowska** [comment]
```
idr-20160803=# select * from annotation_mapvalue where value like '%NIFLUMIC%';
 annotation_id |       name        |                            value                             | index 
---------------+-------------------+--------------------------------------------------------------+-------
       2828971 | Compound Name     | NIFLUMIC ACID                                                |     0
       2828971 | Compound Name URL | https://www.ncbi.nlm.nih.gov/pccompound?term=NIFLUMIC%20ACID |     1
(2 rows)
```

looks good, good to merge

**Comment by joshmoore** [comment]
The general idea of using a proper templating framework for this makes a lot of sense. I am worried about the TODOs, though. Though it's somewhat contradictory, my gut feeling is that these _shouldn't_ be mandatary at runtime while we _should_ require them for tests to pass. More generally, though, this is perhaps a test of the general requirements/virtualenv improvements: if we are sufficiently confident about our deployment strategy, then we can feel more comfortable about introducing Python libraries (whether or not they are `optional_`)

**tl;dr**: for metadata52, this makes good sense. At the latest when its proposed for mainline though, it will need cleaning up, documentation, etc. @manics : do you want to do anything else here before I merge?

**Comment by manics** [comment]
@joshmoore Nothing to add here.

It would be good to get a final policy on python/web requirements though, e.g. for yaml, jinja2, and anything else, and perhaps a standard way of checking for them in code if they're optional?

**Comment by atarkowska** [comment]
that really sounds to me like `pip install omero-metadata`

**Comment by manics** [comment]
Yes, if we resurrect @joshmoore's https://github.com/openmicroscopy/openmicroscopy/pull/4619 PR then all requirements could be handled at the plugin level.

**Comment by joshmoore** [comment]
Merging along with https://github.com/IDR/idr-metadata/pull/169 for 0.3.0-rc1. We can come back to decoupling `populate_metadata` later.

**Comment by atarkowska** [comment]
--rebased-to #5226
